frontend/src/app/services/annotate.service.ts

```diff
@@ -1,34 +1,31 @@
 import { Injectable } from "@angular/core";
-import {
-    catchError,
-    Observable,
-    of,
-} from "rxjs";
+import { catchError, Observable, of, share, Subject, switchMap } from "rxjs";
 import { HttpClient } from "@angular/common/http";
 import { ProblemResponse, ProofBankStats } from "../types";
 
 @Injectable({
     providedIn: "root",
 })
 export class AnnotateService {
-    public problem$(problemId: string): Observable<ProblemResponse | null> {
-        if (!problemId) {
-            console.log("No problem ID provided");
-            return of(null);
-        }
-        return this.http.get<ProblemResponse>(`/api/problem/${problemId}`).pipe(
-            catchError(() => {
-                console.log("Error fetching problem");
-                return of(null);
-            })
-        );
-    }
+    public problemId = new Subject<string>();
```

Comment:
Would something like this work?

```suggestion
    public problem$ = new ReplaySubject<ProblemResponse | null>(1);

    public load(problemId: string) {              
        this.http.get<ProblemResponse>(`/api/problem/${problemId}`).pipe(
            catchError(() => {
                console.error("Error fetching problem");
                return of(null);
            }),
        ).subscribe(this.problem$);
    }

```

Reply:
It probably would, but I expect this create new subscriptions on every `load()` call. Moreover, each subscription will stay active for as long as the application is active. Now in this case, `http.get()` automatically completes once data is received, and you could argue that we want to keep listening for new problems for the app's full lifetime anyway, so in practice these points are a bit moot.

In general, however, I try to subscribe only in components so I can complete the subscriptions when the component is destroyed, and to avoid subscribing in services as much as possible. Component-level services (non-singleton, not provided in 'root') are an exception to this, since they are tightly bound to a component and can also implement `onDestroy` to clear up any subscriptions.